File: zeppelin/src/main/java/org/apache/zeppelin/crossdata/CrossdataInterpreter.java
/*
 * Copyright (C) 2015 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.zeppelin.crossdata;

import com.stratio.crossdata.common.result.ErrorSQLResult;
import com.stratio.crossdata.common.result.SQLResult;
import com.stratio.crossdata.common.result.SuccessfulSQLResult;
import com.stratio.crossdata.driver.JavaDriver;
import com.stratio.crossdata.driver.config.DriverConf;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.types.StructType;
import org.apache.zeppelin.interpreter.Interpreter;
import org.apache.zeppelin.interpreter.InterpreterContext;
import org.apache.zeppelin.interpreter.InterpreterResult;
import org.apache.zeppelin.scheduler.Scheduler;
import org.apache.zeppelin.scheduler.SchedulerFactory;
import scala.concurrent.duration.Duration;
import scala.concurrent.duration.Duration$;
import scala.concurrent.duration.FiniteDuration;

import java.util.Arrays;
import java.util.List;
import java.util.Properties;
import java.util.concurrent.TimeUnit;

public class CrossdataInterpreter extends Interpreter {

    private JavaDriver driver;

    private static final String CROSSDATA_SEEDS_PROPERTY = "crossdata.seeds";
    private static final String CROSSDATA_DEFAULT_LIMIT = "crossdata.defaultLimit";
    private static final String CROSSDATA_TIMEOUT_SEC = "crossdata.timeoutSeconds";


    public CrossdataInterpreter(Properties property) {
        super(property);
    }


    @Override
    public void open() {
        List<String> seeds = Arrays.asList(getProperty(CROSSDATA_SEEDS_PROPERTY).split(","));
        driver = new JavaDriver(new DriverConf().setClusterContactPoint(seeds));
    }

    @Override
    public void close() {
        driver.closeSession();
    }


    @Override
    public InterpreterResult interpret(String sql, InterpreterContext context) {

        long secondsTimeout = Long.parseLong(getProperty(CROSSDATA_TIMEOUT_SEC));

        Duration timeout = (secondsTimeout <= 0) ? Duration$.MODULE$.Inf() : new FiniteDuration(secondsTimeout, TimeUnit.SECONDS);

        SQLResult sqlResult = driver.sql(sql, timeout);

        if (sqlResult.hasError() && ErrorSQLResult.class.isInstance(sqlResult)) {
            return new InterpreterResult(InterpreterResult.Code.ERROR, ErrorSQLResult.class.cast(sqlResult).message());
        } else if (SuccessfulSQLResult.class.isInstance(sqlResult)) {
            StructType schema = SuccessfulSQLResult.class.cast(sqlResult).schema();
            Row[] resultSet = sqlResult.resultSet();
            if (resultSet.length <= 0) {
                return new InterpreterResult(InterpreterResult.Code.SUCCESS, "%text EMPTY result");
            } else {
                return new InterpreterResult(InterpreterResult.Code.SUCCESS, resultToZeppelinMsg(resultSet, schema));
            }

        } else {
            return new InterpreterResult(InterpreterResult.Code.ERROR, "Unexpected result: " + sqlResult.toString());
        }

    }

    private String resultToZeppelinMsg(Row[] resultSet, StructType schema) {

        int defaultLimit = Integer.parseInt(getProperty(CROSSDATA_DEFAULT_LIMIT));

        StringBuilder msg = new StringBuilder();

        // Add columns names
        String resultsHeader = "";
        for (String colName : schema.fieldNames()) {
            if (resultsHeader.isEmpty()) {
                resultsHeader = colName;
            } else {
                resultsHeader += "\t" + colName;
            }
        }

        msg.append(resultsHeader).append(System.lineSeparator());


        //Add rows
        // ArrayType, BinaryType, BooleanType, ByteType, DecimalType, DoubleType, DynamicType,
        // FloatType, FractionalType, IntegerType, IntegralType, LongType, MapType, NativeType,
        // NullType, NumericType, ShortType, StringType, StructType

        int resultLength = Math.min(defaultLimit, resultSet.length);
        int numFields = schema.fieldNames().length;

        for (int r = 0; r < resultLength; r++) {
            Row row = resultSet[r];


            for (int i = 0; i < numFields; i++) {
                if (!row.isNullAt(i)) {
                    msg.append(row.apply(i).toString());
                } else {
                    msg.append("null");
                }
                if (i != numFields - 1) {
                    msg.append("\t");
                }
            }
            msg.append(System.lineSeparator());
        }

        if (resultSet.length > defaultLimit) {
            // TODO use default limit -> Improve driver API
            msg.append(System.lineSeparator()).append("<font color=red>Results are limited by ").append(defaultLimit).append(".</font>");
        }


        msg.append(System.lineSeparator());
        return "%table " + msg.toString();

    }

    @Override
    public void cancel(InterpreterContext context) {
        // TODO do nothing
    }

    @Override
    public FormType getFormType() {
        return FormType.SIMPLE;
    }

    @Override
    public int getProgress(InterpreterContext context) {
        return 0;
    }


    @Override
    public Scheduler getScheduler() {
        return SchedulerFactory.singleton().createOrGetParallelScheduler("interpreter_" + this.hashCode(), 10);
    }
}

File: zeppelin/src/main/scala/org/apache/zeppelin/crossdata/CrossdataInterpreter.scala
package org.apache.zeppelin.crossdata

import java.util.Properties
import java.util.concurrent.TimeUnit

import com.stratio.crossdata.common.result.{ErrorSQLResult, SQLResult, SuccessfulSQLResult}
import com.stratio.crossdata.driver.Driver
import com.stratio.crossdata.driver.config.DriverConf
import org.apache.spark.sql.Row
import org.apache.spark.sql.types.StructType
import org.apache.zeppelin.interpreter.Interpreter.FormType
import org.apache.zeppelin.interpreter.{Interpreter, InterpreterContext, InterpreterResult}
import org.apache.zeppelin.scheduler.{Scheduler, SchedulerFactory}

import scala.collection.JavaConversions._
import scala.concurrent.duration.{Duration, FiniteDuration}

object CrossdataInterpreter {
  private val CROSSDATA_SEEDS_PROPERTY: String = "crossdata.seeds"
  private val CROSSDATA_HTTP_CONNECTION: String = "crossdata.http"
  private val CROSSDATA_DEFAULT_LIMIT: String = "crossdata.defaultLimit"
  private val CROSSDATA_TIMEOUT_SEC: String = "crossdata.timeoutSeconds"
}

class CrossdataInterpreter(property: Properties) extends Interpreter(property) {
  private var driver: Driver = _

  def open() = {
    val seeds: java.util.List[String] = (getProperty(CrossdataInterpreter.CROSSDATA_SEEDS_PROPERTY).split(",")).toList
    val driverConf = new DriverConf

    driver = if(getProperty(CrossdataInterpreter.CROSSDATA_HTTP_CONNECTION).toBoolean){
      seeds(0).split(":") match { case Array(host, port) => driverConf.setHttpHostAndPort(host, port.toInt) }
      Driver.http.newSession(driverConf)
    } else {
      Driver.newSession(seeds, driverConf)
    }

  }

  def close() = {
    driver.closeSession
  }

  def interpret(sql: String, context: InterpreterContext): InterpreterResult = {

    val secondsTimeout: Long = (getProperty(CrossdataInterpreter.CROSSDATA_TIMEOUT_SEC)).toLong

    val timeout: Duration = if (secondsTimeout <= 0) Duration.Inf else new FiniteDuration(secondsTimeout, TimeUnit.SECONDS)

    driver.sql(sql).waitForResult(timeout) match {
      case er: ErrorSQLResult => new InterpreterResult(InterpreterResult.Code.ERROR, er.message)
      case esr: SuccessfulSQLResult if esr.resultSet.length <= 0 => new InterpreterResult(InterpreterResult.Code.SUCCESS, "%text EMPTY result")
      case sr: SuccessfulSQLResult if sr.resultSet.length > 0 => new InterpreterResult(InterpreterResult.Code.SUCCESS, resultToZeppelinMsg(sr))
      case other: SQLResult => new InterpreterResult(InterpreterResult.Code.ERROR, s"Unexpected result: ${other.toString}")
    }
  }

  private def resultToZeppelinMsg(sqlResult: SQLResult): String = {
    resultToZeppelinMsg(sqlResult.resultSet, sqlResult.schema)
  }

  private def resultToZeppelinMsg(resultSet: Array[Row], schema: StructType): String = {
    val defaultLimit: Int = getProperty(CrossdataInterpreter.CROSSDATA_DEFAULT_LIMIT).toInt
    val msg: StringBuilder = new StringBuilder
    // Add columns names

    msg.append(schema.fieldNames.mkString("\t")).append(System.lineSeparator)

    //Add rows
    // ArrayType, BinaryType, BooleanType, ByteType, DecimalType, DoubleType, DynamicType,
    // FloatType, FractionalType, IntegerType, IntegralType, LongType, MapType, NativeType,
    // NullType, NumericType, ShortType, StringType, StructType
    resultSet.take(Math.min(defaultLimit, resultSet.length)) foreach { r =>
      val arrayRow: Seq[String] = for (i <- 0 until r.length) yield {
        if(r.isNullAt(i)) "null" else r(i).toString
      }
      msg.append(s"${arrayRow.mkString("\t")}${System.lineSeparator}")
    }

    if (resultSet.length > defaultLimit) {
      // TODO use default limit -> Improve driver API
      msg.append(System.lineSeparator).append("<font color=red>Results are limited by ").append(defaultLimit).append(".</font>")
    }

    s"%table ${msg.append(System.lineSeparator).toString}"
  }

  def cancel(context: InterpreterContext) = {
    // TODO do nothing
  }

  def getFormType: Interpreter.FormType = FormType.SIMPLE

  def getProgress(context: InterpreterContext): Int = 0

  override def getScheduler: Scheduler =
    SchedulerFactory.singleton.createOrGetParallelScheduler(s"interpreter_${this.hashCode}", 10)
}

